### PR TITLE
Fix issues related to delay when processing events on X11 display server

### LINKED ISSUE #31177: Keyboard repeat lag on GNOME
**Godot version:**
3.1.1 stable, also happens on 3.2 dev

**OS/device including version:**
OS: Linux, Fedora 30, GNOME 3.32.2
GPU: Intel UHD 620

**Issue description:**
When using the script editor on GNOME desktop, the response to keyboard repeat is “lagging”.
It’s difficult to illustrate with a GIF, but simply put: if you hold a key (e.g. arrow down to scroll) and then release it, the release triggers not immediately but after a certain lag.
Response to initial keystroke is immediate, so the lag only happens on keyboard repeat, i.e. when holding a key.

This happens only on GNOME, both when running on Wayland and Xorg.
I’ve tested on i3 and Sway: the problem disappears.

Adjusting keyboard repeat rate in GNOME also affects this issue but doesn’t solve it: higher repeat rate makes the lag in Godot longer. So if I have a very high repeat rate, Godot might keep “firing” the repeated keystroke for over ~2sec after I release the key.

To exclude the possibility that it's specific to my machine and OS setup, I’ve also tested and confirmed the issue on a spare laptop with a clean install of Fedora 29.

### LINKED ISSUE #39422: Xlib clipboard/selection requests do not play well with the powersave FPS-cap
**Godot version:**
Godot Engine v3.2..2.beta.custom_build.aa57bb047
aa57bb0473cb3930670bcfa3159f96262dfaa1dd

Edit: kind of a regression (see https://github.com/godotengine/godot/issues/39422#issuecomment-645950551 ). Introduced in ac14efcdfb22575fc49a75201492955e723d04c3. 

Or instead of decoupling, make the part that deals with selections and clipboards more efficient somehow? Xlib is not very enjoyable to deal with.

**OS/device including version:**
Linux

**Issue description:**
* Firefox with autoscroll enabled (pressing and holding middle click and dragging the mouse up causes page to scroll)
* press ctrl+c in the editor in the client. Does not need to have text selected
* Try to hold+drag middle in firefox, then after a few seconds release it
For me, the page in firefox keeps scrolling for around a second before stopping scrolling.
This issue persists until I select some text in firefox. After that, scrolling stops immediately upon release again.

**Steps to reproduce:**
1) Have the editor in a project and Firefox (my version is 76) open
2) Press ctrl+c in the editor
3) Press and hold middleclick in firefox, move the mouse a bit, release middleclick
The reticle lags behind for a second or so, long after the middleclick was released
In order to "fix it", select some text in Firefox. This fixes it until the above steps are repeated again.

### PR DESCRIPTION
**Pull request by pouleyKetchoupp**
*Edit 09/25/2020: Second commit added to fix clipboard issues.* 

This change moves the handling of events from x server to a separate thread to fix specific issues due to the delay caused by processing them in the main loop, which didn't allow for asynchronous communication with the x server.

**Fix general keyboard input lag on X11 display server**
The first commit makes keyboard inputs more responsive on Linux, especially when the FPS is lower on slower configurations.

Polling events from the x server is done on a separate thread to avoid a frame delay with inputs, due to first sending the event to the input manager with `XFilterEvent` then processing the new event only on the next frame.

Calls to Input Manager functions like `XSetICFocus`, `XUnsetICFocus` and `XSetICValues` use a mutex, because they are polling events internally and would otherwise interfere with our own thread process for polling events which can cause a deadlock in some cases.

`XUnsetICFocus` is called instead of `XSetICFocus` on `FocusOut` events, so the input manager can be properly notified of focus changes.

`clipboard_get` now uses a blocking call to poll for a specific event type when waiting for a `SelectionNotify` event, instead of polling all events and filtering them afterwards.

Fixes #31177

**Fix delay to process clipboard content from Godot in other programs**
The second commit adds extra changes to fix issues with clipboard handling.

When pasting clipboard content from Godot to other applications, multiple SelectionRequest events are sent to Godot in order to access the data. It could take a long time before the data is ready for the other app because events were processed one by one on the main thread, especially when Godot is unfocused and runs at low frequency.

With this change, SelectionRequest events are directly handled on the separate event polling thread to minimize this delay.

This change also replaces clipboard_get() calls in SelectionRequest with a direct access to internal_clipboard, since in this case we know Godot is the owner of the clipboard content and it's not necessary to query the x server for it.

Fixes #39422

### CODE REVIEW

File: platform/linuxbsd/display_server_x11.cpp [hunk 1]
```diff
@@ -2325,6 +2345,56 @@ void DisplayServerX11::_send_window_event(const WindowData &wd, WindowEvent p_ev
...
+		int num_ready_fds = select(x11_fd + 1, &in_fds, NULL, NULL, &tv);
+		if (num_ready_fds < 0) {
+			ERR_PRINT("_poll_events: select error: " + itos(errno));
+		}
+
+		// Process events from the queue.
+		{
+			MutexLock mutex_lock(events_mutex);
+
+			auto predicate = [](Display *display, XEvent *event, XPointer arg) -> Bool {
```
Comment (reduz) [inline platform/linuxbsd/display_server_x11.cpp:2377]: Please use a separate function and don't use auto, I know its a bit more of a hassle, but we aim at very explicit C++ code style.

File: platform/linuxbsd/display_server_x11.cpp [hunk 2]
```diff
@@ -2325,6 +2345,56 @@ void DisplayServerX11::_send_window_event(const WindowData &wd, WindowEvent p_ev
...
+				if (XFilterEvent(&ev, None)) {
+					// Event has been filtered by the Input Manager,
+					// it has to be ignored and a new one will be received.
+					continue;
+				}
+				polled_events.push_back(ev);
+			}
+		}
+	}
+}
```
Comment (reduz) [inline platform/linuxbsd/display_server_x11.cpp:2396]: Do you not need some sort of locking to protect from calls to x11_display in the main thread? like for setting window size and other stuff

**Review by reduz — CHANGES_REQUESTED** [review]

File: platform/linuxbsd/display_server_x11.cpp (hunk 2 above)
Comment (pouleyKetchoupp) [reply platform/linuxbsd/display_server_x11.cpp:2396]: XLib calls generally use XLockDisplay/XUnlockDisplay internally, so it should be safe as long as they don't poll events from the event queue while our thread process is doing it as well. I've used our own mutex when this case happens.

Comment (pouleyKetchoupp) [reply platform/linuxbsd/display_server_x11.cpp:2396]: Maybe @hpvb can validate that I'm not overlooking anything about thread safety.

File: platform/linuxbsd/display_server_x11.cpp (hunk 1 above)
Comment (pouleyKetchoupp) [reply platform/linuxbsd/display_server_x11.cpp]: Fixed.

**Comment by pouleyKetchoupp** [comment]
Just pushed an extra commit to fix a issue with clipboard contents on top of the delay with keyboard inputs, since both of these changes need events to be processed on a separate thread.

@hpvb I know you have a plan for refactoring x11 events processing, but I figured that for now it doesn't seem too difficult to selectively process certain events on the separate thread to fix specific issues. It should be even possible to backport these changes to the 3.2 branch, and it's possible to refactor more things later for 4.0.

**Comment by akien-mga** [comment]
Thanks! (Note: approved on IRC by TMM)

**Pull request merged**